File: ai/mcp/typescript/McpHost.ts
import { McpClient } from './McpClient.js'
import { McpServerBase } from './McpServerBase.js'
import { McpTool, McpFunctionTool } from './McpTypes.js'

/**
 * mcp client model.
 */
export interface McpClientModel {
    id: string;
    client: McpClient;
}

/**
 * mcp server model.
 */
export interface McpServerModel {
    id: string;
    server: McpServerBase;
}

/**
 * Model context protocol host.
 */
export class McpHost {

    // global.
    private mcpClients: Array<McpClientModel>;
    private mcpServers: Array<McpServerModel>;
    private mcpFunctionTools: Array<McpFunctionTool>;

    // event logging function.
    private logEvent: (eventType: string, name: string, message: string, details: any) => void;

    /**
     * Model context protocol host.
     */
    constructor() {

        // init
        this.mcpClients = [];
        this.mcpServers = [];
        this.mcpFunctionTools = [];
        this.logEvent = null;
    }

    /**
     * Subscribe to the on event.
     *
     * @param {function}	event callback(eventType, name, message, details).
     */
    onEvent(event: (eventType: string, name: string, message: string, details: any) => void): void {
        // assign the event.
        this.logEvent = event;
    }

    /**
     * get the list of clients
     * @returns {Array<McpClientModel>} list of clients
     */
    getClients(): Array<McpClientModel> {
        return this.mcpClients;
    }

    /**
     * get the list of servers
     * @returns {Array<McpServerModel>} list of servers
     */
    getServers(): Array<McpServerModel> {
        return this.mcpServers;
    }

    /**
     * get the list of function tools
     * @returns {Array<McpFunctionTool>} list of function tools
     */
    getFunctionTools(): Array<McpFunctionTool> {
        return this.mcpFunctionTools;
    }

    /**
     * add an MCP client.
     * @param {string} id   the unique id.
     * @param {McpClient} client    the mcp client.
     */
    addClient(id: string, client: McpClient): void {
        // add a new client.
        this.mcpClients.push({
            id: id,
            client: client
        });
    }

    /**
     * add an MCP server.
     * @param {string} id   the unique id.
     * @param {McpServerBase} server    the mcp server.
     */
    addServer(id: string, server: McpServerBase): void {
        // add a new server.
        this.mcpServers.push({
            id: id,
            server: server
        });
    }

    /**
     * add an MCP function tool.
     * @param {McpTool} tool    the mcp tool.
     * @param {string} clientId    the mcp client Id.
     * @param {string} serverId    the mcp server Id.
     */
    addFunctionTool(tool: McpTool, clientId: string, serverId: string): void {
        // add a new function tool.
        this.mcpFunctionTools.push({
            clientId: clientId,
            serverId: serverId,
            type: "function",
            name: tool.name,
            description: tool.description,
            strict: true,
            inputSchema: tool.inputSchema,
            parameters: tool.parameters
        });
    }

    /**
     * find an MCP client.
     * @param {string} id   the unique id.
     * @returns {McpClientModel} the mcp client
     */
    findClient(id: string): McpClientModel | null {
        return this.mcpClients.find(function (item) {
            return (item.id === id || item.id === id);
        })
    }

    /**
     * find an MCP server.
     * @param {string} id   the unique id.
     * @returns {McpServerModel} the mcp server
     */
    findServer(id: string): McpServerModel | null {
        return this.mcpServers.find(function (item) {
            return (item.id === id || item.id === id);
        })
    }

    /**
     * find an MCP function tool.
     * @param {string} name   the unique name.
     * @returns {McpFunctionTool} the mcp function tool
     */
    findFunctionTool(name: string): McpFunctionTool | null {
        return this.mcpFunctionTools.find(function (item) {
            return (item.name === name || item.name === name);
        })
    }

    /**
    * close all clients and servers.
    */
    async closeAll(): Promise<void> {
        await this.closeServers();
        await this.closeClients();

        this.mcpClients = [];
        this.mcpServers = [];
        this.mcpFunctionTools = [];

        this.logEvent = null;
    }

    /**
    * close all servers.
    */
    async closeServers(): Promise<void> {

        // for each mcp server.
        for (var i = 0; i < this.mcpServers.length; i++) {
            try {
                // close.
                await this.mcpServers[i].server.stopServer();
            } catch (e) {
                if (this.logEvent) this.logEvent("error", "close", "close server", e);
            }
        }
    }

    /**
    * close all clients.
    */
    async closeClients(): Promise<void> {

        // for each mcp client.
        for (var i = 0; i < this.mcpClients.length; i++) {
            try {
                // close.
                await this.mcpClients[i].client.closeConnection();
            } catch (e) {
                if (this.logEvent) this.logEvent("error", "close", "close client", e);
            }
        }
    }

    /**
     * add the server function tools.
     * @param {string} id     the unique id.
     * @param {McpServerBase} mcpServer the mcp server
     */
    addServerFunctionTools(id: string, mcpServer: McpServerBase): void {

        // add each server.
        this.addServer(id, mcpServer);

        // add tools.
        let tools: McpTool[] = mcpServer.getTools();

        // assign function
        for (var i = 0; i < tools.length; i++) {
            // add to tools
            let tool: McpTool = tools[i];
            this.addFunctionTool(tool, "", id);
        }
    }

    /**
    * add the client function tools.
    * @param {string} id     the unique id.
    * @param {McpClient} mcpClient the mcp client
    */
    addClientFunctionTools(id: string, mcpClient: McpClient): void {

        // add each client.
        this.addClient(id, mcpClient);

        // add tools.
        let tools: McpTool[] = mcpClient.getTools();

        // assign function
        for (var i = 0; i < tools.length; i++) {
            // add to tools
            let tool: McpTool = tools[i];
            this.addFunctionTool(tool, id, "");
        }
    }
}

File: ai/mcp/python/McpHost.py
from typing import Optional, Any, List, Union, Dict

from .McpClient import McpClient
from .McpServerBase import McpServerBase
from .McpTypes import McpTool, McpFunctionTool

# Model context protocol client model.
class McpClientModel:
    """
    Model context protocol client model.
    """
    def __init__(self,
                 id: str,
                 client: McpClient):
        self.id = id
        self.client = client

    def __repr__(self):
        return f"McpClientModel(id={self.id}, " \
            f"client={self.client})"

# Model context protocol server model.
class McpServerModel:
    """
    Model context protocol server model.
    """
    def __init__(self,
                 id: str,
                 server: McpServerBase):
        self.id = id
        self.server = server

    def __repr__(self):
        return f"McpServerModel(id={self.id}, " \
            f"server={self.server})"

# Model context protocol host.
class McpHost:
    """
    Model context protocol host.
    """
    def __init__(self):
        # init
        self.mcpClients: List[McpClientModel] = [];
        self.mcpServers: List[McpServerModel] = [];
        self.mcpFunctionTools: List[McpFunctionTool] = [];

    def getClients(self) -> List[McpClientModel]:
        """
        get the list of clients

        Return:
            list of clients
        """
        return self.mcpClients

    def getServers(self) -> List[McpServerModel]:
        """
        get the list of servers

        Return:
            list of servers
        """
        return self.mcpServers

    def getFunctionTools(self) -> List[McpFunctionTool]:
        """
        get the list of function tools

        Return:
            list of function tools
        """
        return self.mcpFunctionTools

    def addClient(self, id: str, client: McpClient):
        """
        add an MCP client.

        Args:
            id: the unique id.
            client: the mcp client.
        """
        self.mcpClients.append(McpClientModel(
                id,
                client
            ))

    def addServer(self, id: str, server: McpServerBase):
        """
        add an MCP server.

        Args:
            id: the unique id.
            server: the mcp server.
        """
        self.mcpServers.append(McpServerModel(
                id,
                server
            ))

    def addFunctionTool(self, tool: McpTool, clientId: str, serverId: str):
        """
        add an MCP function tool.

        Args:
            tool: the mcp tool.
            clientId: the mcp client Id.
            serverId: the mcp server Id.
        """
        self.mcpFunctionTools.append(McpFunctionTool(
                clientId,
                serverId,
                "function",
                tool.name,
                tool.description,
                True,
                tool.inputSchema,
                tool.parameters
            ))

    def findClient(self, id: str) -> McpClientModel | None:
        """
        find an MCP client.

        Args:
            id: the unique id.

        Return:
            the mcp client; else empty
        """
        found: bool = False
        clientModel: McpClientModel = None

        # for each
        for client in self.mcpClients:
            if id == client.id:
                found = True
                clientModel = client
                break

        # return
        return clientModel

    def findServer(self, id: str) -> McpServerModel | None:
        """
        find an MCP server.

        Args:
            id: the unique id.

        Return:
            the mcp server; else empty
        """
        found: bool = False
        serverModel: McpServerModel = None

        # for each
        for server in self.mcpServers:
            if id == server.id:
                found = True
                serverModel = server
                break

        # return
        return serverModel

    def findFunctionTool(self, name: str) -> McpFunctionTool | None:
        """
        find an MCP function tool.

        Args:
            name: the unique name.

        Return:
            the mcp function tool; else empty
        """
        found: bool = False
        functionTool: McpFunctionTool = None

        # for each
        for tool in self.mcpFunctionTools:
            if name == tool.name:
                found = True
                functionTool = tool
                break

        # return
        return functionTool